lib/AST/ClangTypeConverter.cpp

```diff
@@ -0,0 +1,688 @@
+      // Only tuples where all element types are equal map to fixed-size
+      // arrays.
+      return clang::QualType();
+  }
+
+  auto clangEltTy = convert(eltTy);
+  if (clangEltTy.isNull())
+    return clang::QualType();
+
+  APInt size(32, tupleNumElements);
+  return ClangASTContext.getConstantArrayType(clangEltTy, size,
+           clang::ArrayType::Normal, 0);
+}
+
+clang::QualType ClangTypeConverter::visitProtocolType(ProtocolType *type) {
+  auto proto = type->getDecl();
+
+  // Single protocol -> id<Proto>
+  if (proto->isObjC()) {
+    auto &clangCtx = ClangASTContext;
+    clang::IdentifierInfo *name = &clangCtx.Idents.get(proto->getName().get());
+    auto *PDecl = clang::ObjCProtocolDecl::Create(
+                    const_cast<clang::ASTContext &>(clangCtx),
+                    clangCtx.getTranslationUnitDecl(), name,
+                    clang::SourceLocation(), clang::SourceLocation(), nullptr);
+
+    // Attach an objc_runtime_name attribute with the Objective-C name to use
+    // for this protocol.
+    SmallString<64> runtimeNameBuffer;
+    PDecl->addAttr(clang::ObjCRuntimeNameAttr::CreateImplicit(
+                     PDecl->getASTContext(),
+                     proto->getObjCRuntimeName(runtimeNameBuffer)));
+
+    auto clangType  = clangCtx.getObjCObjectType(clangCtx.ObjCBuiltinIdTy,
+                                                 &PDecl, 1);
+    return clangCtx.getObjCObjectPointerType(clangType);
+  }
+
+  return getClangIdType(ClangASTContext);
+}
+
+clang::QualType ClangTypeConverter::visitMetatypeType(MetatypeType *type) {
+  return getClangMetatypeType(ClangASTContext);
+}
+
+clang::QualType
+ClangTypeConverter::visitExistentialMetatypeType(ExistentialMetatypeType *type) {
+  return getClangMetatypeType(ClangASTContext);
+}
+
+clang::QualType ClangTypeConverter::visitClassType(ClassType *type) {
+  auto &clangCtx = ClangASTContext;
+  // produce the clang type INTF * if it is imported ObjC object.
+  auto swiftDecl = type->getDecl();
+  if (swiftDecl->isObjC()) {
+    clang::IdentifierInfo *ForwardClassId =
+      &clangCtx.Idents.get(swiftDecl->getName().get());
+    auto *CDecl = clang::ObjCInterfaceDecl::Create(
+                          clangCtx, clangCtx.getTranslationUnitDecl(),
+                          clang::SourceLocation(), ForwardClassId,
+                          /*typeParamList*/nullptr, /*PrevDecl=*/nullptr,
+                          clang::SourceLocation());
+
+    // Attach an objc_runtime_name attribute with the Objective-C name to use
+    // for this class.
+    SmallString<64> runtimeNameBuffer;
+    CDecl->addAttr(clang::ObjCRuntimeNameAttr::CreateImplicit(
+                     CDecl->getASTContext(),
+                     swiftDecl->getObjCRuntimeName(runtimeNameBuffer)));
+
+    auto clangType  = clangCtx.getObjCInterfaceType(CDecl);
+    return clangCtx.getObjCObjectPointerType(clangType);
+  }
+  return getClangIdType(clangCtx);
```

Comment:
@rjmccall I'm guessing this has the same issues as with the `visitProtocolType` method? (Not caching decls and returning a bogus `id` for non-ObjC stuff)

Reply:
Github lost track of where this is after the rebase.  If this is the protocol-composition method, yeah, ideally they would re-use a common method for translating a `ProtocolDecl`.  If this is the class method, then yeah, same principle: we should cache the "exported" `ObjCInterfaceDecl`.

Reply:
Yes, this is in the `visitClassType` method. If I error out in `GenClangType` (which has similar code) if `!swiftDecl->isObjC()`, then I get a crash with

```
class SwiftClass {
      @objc func objcmethod() -> () {}
      func swiftmethod() -> () {}
      // Block attribute
      func f(_ someBlock: @convention(block) (Int64) -> Int64) {
      }
}
```

Not entirely sure why we're trying to convert that.